### ISSUE TITLE
Prepull custom images

### ISSUE THREAD

**Issue by paalkr** [kubernetes-incubator/kube-aws#571]
Hi

I have some quite large docker images (around 7GB in size), and I would suggest an option to prepull custom images during node boot so that the first deploy of a pod will be quick. I'm fully aware that this would lead to a longer boot time for a node. A node should not be reported as ready before all custom images are pulled completely. A config in the cluster.yaml file might look like

```
worker:
  nodePools:
    - name: poolname
      prePullImages:
        - "repo:label"
        - "repo2:label"
```
Pulling from a private repo like AWS ECR should be supported.

Any thought on this feature request?

PS, I understand that there has been some discussion regarding prepull of images vs baking custom AMIs.

**Comment by redbaron** [comment]
republish to ecr?

**Comment by paalkr** [comment]
@redbaron . Sorry, I'm not sure I understand what you mean by republish to ecr?

**Comment by c-knowles** [comment]
Are these custom images running cluster services or your own apps? If it's your own apps, what is the use case to preload rather than waiting on app deploy/update? Helm for example waits for deploys to happen so just trying to figure out what the intention is here.

**Comment by paalkr** [comment]
These are custom images driving my own services and workloads. I will try to explain the motivation and use case.

Pulling a 7GB custom image from ECR to a node takes approx 4-5 minutes Including unpack), slightly dependent on the node type. Imagine that I have a service (pods) that is scaled by a horizontal pod autoscaler rule, and by the time the hpa would like to add one pod for the service, a new node was just added to the pool (by node auto scaling). The pod scheduler would then try to put the pod on the newly started node, and this will lead to a 4-5 minutes delay for the pod to become ready (due to image pulling delay).

So in this case it would actually been better if the new pod was scheduled to a existing node, already having the image, to quickly increase the service capacity.

Or as I suggest the image could be prepulled during node startup, and then the service capacity would be increased quickly regardless of which node the new pod is scheduled to run.

Does this make sense?

**Comment by c-knowles** [comment]
I think there are these cases for up scaling:
1. Lack of resources which means cluster scales up
  * if preloading it won't take any time off scaling because the instance still has to download the images
  * without preloaded the instances start up quicker
1. New pod fits on one of the existing instances
 * if preloaded it means the pod can start quicker but if not and we are scaling up and down a lot all instances eventually naturally end up with the image (at least in the short term, see below caveat)
  * if not preloaded it means the image takes some time to download

I'm not sure in 2 if the scheduler is clever enough to launch pods on instances which had the image before. The big caveat is that images can be garbage collected by k8s so without a pod using the image I believe it can get deleted.

**Comment by redbaron** [comment]
IMHO running apps is not a concern of kube-aws. Are you going to redeploy your cluster when your all image updates? It doesn’t make sense.
 
 
As a hack You can solve it by creating your own pre-puller in the form of DaemonSet which runs your image, but not the app. Let's say you override entrypoint to be /bin/bash -c ‘sleep infinity’
 
Real solution would be:
Not to have 7GB images. When you build image do install and remove unnecessary bits in the same RUN command, if you separate them, they are committed to different layers and total image size doesn’t shrink.
If you can not avoid 7GB images, try to move RUN commands which create most of data to the top, and overall move most frequently changed parts to the bottom and least frequently changed paths to the top. This will help with layers reuse, so that when new version of the image is pulled, only small number of layers are actually downloaded.

**Comment by paalkr** [comment]
Thanks for the feedback

@redbaron , I'm not going to reload my cluster for a service update. The image I would like to prepull is a "base image". Changes to services will be applied as a small additional image layer of only a few megabytes.

The base image is highly optimized when i comes to size by concatenating RUN commands in the dockerfile. Also download, install and removal of any intermediate install files are concatenated to avoid extra layers in the image.

The use case of ours is maybe a little special. We need to use the AWS Auto Scaling features to scale our cluster worker pools, the Kubernetes cluster autoscaler doesn't work well with our use case. I will try to explain why.

When I deploy a service pod I would for the most times use a very small CPU request in the pod sepc, maybe as small as 100m CPU. Many of the several hundreds of services we intend to deploy to our cluster, has a very low request rate (hence low idle load). But when a service gets a request the pod will usually consume a complete host core (or whatever I cap the pod at with CPU resource limit) for as long as it takes to process the request (can be milliseconds to seconds). Once in a while one or some of the services get a dramatic increase in load that might last for hours, that lead to a considerable horizontally scaled out pod.

So it does not make sense to us to request say 500m CPU for each service pod, usually they just stay there idle for 90% of the day consuming not CPU at all on the underlying host. So we in practice overbook the host CPUs dramatically. The Kubernetes cluster autoscaler will not understand that some of the services are under heavy load, as the real node load is not used to calculate the need for new nodes. Kubernetes will only add a node if it don't find room to add a pod to any nodes based on the pod resource request (which in our case is way smaller then the real load, when the service actually have to do work). It's a calculated risk that not all services get under heavy load at the same time ;)

So by using AWS Auto Scaling we can scale out the nodes based on actual node CPU load, regardless if a node theoretically might fit more pods according to the sum of pod CPU requests for pr node. This, in combination with podAntiAffinity and preferredDuringSchedulingIgnoredDuringExecution, will usually result in new pods landing on any newly added nodes, if hpa decides to scale out pod.  Which without image prepull will not have the base image in place.

Ideally the Kubernetes cluster autoscaler should look at both sum of resource requests AND real load on nodes when deciding if a pool needs to scale out with more nodes. And the scheduler should look at node loads not only if the node theoretically have room for a pod (based on pod resource requests), when deciding which node to put a new pod.

I'm not sure if I'm any of this make sense. It would be easier to explain this by speaking, and maybe a demo. ;)

**Comment by mumoshu** [comment]
@paalkr Thanks for the detailed feedback 👍 

First of all, I agree with you that the AWS native auto scaling fits your use-case.

However, on the other hand, I guess you can still utilize k8s' cluster-autoscaler in combination with horizontal-pod-autoscaler(HPA) and pod anti affinity instead.
My idea is that you set cpu request to a minimum amount enough for idling, say 100m, and enable HPA on your replica set to trigger scaling with a threshold of, say, 90% of the requested amount 100m. Also don't forget to provide a pod anti affinity not to make pods in the replica set collocated into a node. This way, once pods in your replica set start consuming cpu more than requested 100m, hpa will auto-scales pods into different nodes, which in turn triggers cluster-autoscaler to add more nodes.

Also, my guess is that utilizing the AWS native auto scaling here would still require you to enable HPA anyway, or your replicas will end up at the same number. However, if you enable HPA in the first place, cluster-autoscaler would work IMHO.

---

Regarding your original request of pre-pulling images, my preference is "baking them into AMIs"
 #505. So that we don't need to wait until cluster-autoscaler is improved to allow having "spare capacity" for faster pod autoscaling https://github.com/kubernetes/contrib/issues/2251.
In other words, you'd need to wait for the CA improvement to achieve rapid auto-scaling of pods with your possible Low-CPU-Request/HPA/Pod-Anti-Affinity/Cluster-Autoscaler/Image-Pre-Pulling setup.

What do you think?

**Comment by paalkr** [comment]
Hi @mumoshu ,

This is complicated, that's for sure, with many moving parts in actions ;) We do relay on HPA for scaling our pods, which works quite well. The "problem" for us handling the scaling out for more nodes when some pods get under heavy load. If you sum the requested amount of CPU pr node, there wouldn't be no need for adding more nodes to the cluster. But if you sum the real CPU load of the cluster nodes you would see that real CPU-load is so high at all nodes that more nodes definitely should be added to the pool.

The nice thing is that you will see the node real load by running `top nodes`. But unfortunately this information is not evaluated by the K8s cluster autoscaler, if I'm not mistaken. K8s cluster autoscaler only operates by the sum of resource requests pr node, not the nodes actual CPU load.

An AWS CloudWatch alarm hooked up to our ASG CPU utilization will on the other hand kick in, and scale the pool if needed based on real node CPU metrics. For the time being we're quite happy with how AWS ASG auto scaling works though. Especially as we are using spot fleet for some worker pool. Spot Fleet capacity can sins a while ago (https://aws.amazon.com/blogs/aws/new-auto-scaling-for-ec2-spot-fleets/) be adjusted (AutoScaled) based on CPU metrics for the fleet, without the fleet being recreated.

**Comment by mumoshu** [comment]
Hi @paalkr, thanks as always for the detailed reply.

> If you sum the requested amount of CPU pr node, there wouldn't be no need for adding more nodes to the cluster.

This is becoming more and more my education but then, how do you scale your PODS after the AWS native autoscaling added NODES?

**Comment by paalkr** [comment]
We do set the AWS CPU AutoScaling threshold at a level that will ensure the pool scales out say at 35% real ASG CPU load. This will make sure we do have some free capacity in the worker pool (for any service/deployment to scale within normal load). We also assumes (by historical data) that not all deployments/services will be under heavy load at the same time. So by using the 35% threshold the worker pool will almost always be in advance of the pods scaling, and have free capacity when some pods get a spike in load.

HPA evaluates the real pod CPU utilization, and decides to scale out the pods if a deployment/service get under heavy load. The pod anti affinity will try to spread the load on all nodes. As the ASG threshold is breached more capacity/nodes are added to the worker pool. Any new pods will then end up at these new nodes due to the fact that we have pod anti affinity and that the scheduler places new pods on nodes with the most free resources (it least I'm under the impression that the scheduler behaves like this). And in this context having the base image for our pods/service in place (pre-pulled) will improve the speed of pod auto scaling/initial startup time.

I'm not sure if my English is good enough to precisely describe the use case, but I'm trying my best ;)

**Comment by mumoshu** [comment]
@paalkr Thanks again for the reply.

> and have free capacity when some pods get a spike in load.

So basically,
* You're using AWS-native autoscaling instead of cluster-autoscaler just for adding nodes earlier(in advance/before a spike comes)
* cluster-autoscaler can be used for your use-case onc https://github.com/kubernetes/contrib/issues/2251 is resolved

right?

> I'm not sure if my English is good enough to precisely describe the use case, but I'm trying my best ;)

So am I :)

Revisiting your original question, for me [@redbaron's idea of creating a pre-puller daemonset](https://github.com/kubernetes-incubator/kube-aws/issues/571#issuecomment-296131538) seems to work for you today.

I guess we can even support automating deployments of the pre-puller daemonset once #509 comes to reality.

What do you think?

**Comment by paalkr** [comment]
@mumoshu , thanks :)

>So basically,
You're using AWS-native autoscaling instead of cluster-autoscaler just for adding nodes earlier(in advance/before a spike comes)
cluster-autoscaler can be used for your use-case onc kubernetes/contrib#2251 is resolved
right?

Probably
>Revisiting your original question, for me @redbaron's idea of creating a pre-puller daemonset seems to work for you today.

The idea of a DeamonSet to pull the image is maybe not the optimal solution, as the node will be rendered "ready" before the pull process are finished.

Plugin seems like a feature that we might be able to use to get our desired functionality implemented.

**Comment by mumoshu** [comment]
Thanks @paalkr!

> as the node will be rendered "ready" before the pull process are finished.

Good point 👍 It must be true.
I wish I could add a generic mechanism to "inject" your custom dependencies (backed by systemd unit, ExecStartPre of a systemd service, k8s job targeted to the node?, etc.) which delays a node from becoming "ready" until those dependencies are met.

## Option 1

```yaml
worker:
  nodePools:
    - name: poolname
      prePullImages:
        - "repo:label"
        - "repo2:label"
```

This is your original idea which would work very well with this specific use-case

## Option 2

```yaml
worker:
  nodePools:
    - name: poolname
      prewarm:
      - command: bash
        args:
        - -c
        - |
          docker pull repo:label
          docker pull repo2:label
```

would be a bit more generic.

**Comment by paalkr** [comment]
@mumoshu , thanks!

Any of the options would work as I see it, as long as custom repo credentials (ECR AIM or any other private repo logins) applied to a node is respected, so the docker pull isn't blocked.

Your suggestion is more generic then mine, and can probably be used to solve other custom needs as well.

**Comment by redbaron** [comment]
@mumoshu , @paalkr don't we have custom systemd unit files support now? there can be a unit file which does `docker pull`, if you make it `RequiredBy=kubelet.service` then then node will be joining kube with image pre-pulled

**Comment by mumoshu** [comment]
@redbaron @paalkr Wow, thanks. I've been completely forgotten about the feature...
Untested but I guess something like this would work

```
worker:
  nodePools:
  - name: pool1
     customSystemdUnits:
     - name: prepull.service
       command: start
       enable: true
       content: |
         [Unit]
         Description=Docker images pre-puller
         RequiredBy=kubelet.service
         [Service]
         Type=oneshot
         RemainAfterExit=yes
         ExecStartPre=/bin/bash -c "for img in repo:label repo2:label; do echo pulling \$img; until /usr/bin/docker pull \$img; do echo retrying...; sleep 1; done; done; echo successfully pulled all the images"
         ExecStart=/bin/true
```

**Comment by paalkr** [comment]
@redbaron @mumoshu 

Thanks! Clever suggestion. I'd never come to think that I could use a custom Systemd service to prepull the images. I will try out this suggestion and report back.

**Comment by mumoshu** [comment]
Hi @paalkr, did it work for you? 😄

**Comment by paalkr** [comment]
Hi @mumoshu . Not during my first try. I have to look at the command and see if I made any typos.

**Comment by paalkr** [comment]
Hi @mumoshu 

The service failed as you can see from the screen shot
![image](https://cloud.githubusercontent.com/assets/9726307/26138960/6b116bea-3acd-11e7-8f11-3d331718add5.png)

Can you please direct me to where the log file is stored?

**Comment by paalkr** [comment]
@mumoshu ,

BTW, I understand why the kube-node-label.service failed in this particular case. The above screen shot is from a spot fleet node, and the kube-node-label.service is referring the AWS resources (like auto scaling group and launch configuration) which is not available for a spot fleet ;)

**Comment by mumoshu** [comment]
Hi @paalkr, I guess you can `journalctl -u prepull.service > prepull.service.log` to save full log of the service.

**Comment by mumoshu** [comment]
@paalkr 

> kube-node-label.service is referring the AWS resources (like auto scaling group and launch configuration) which is not available for a spot fleet ;)

Thanks for the good catch!
Probably we're missing a validation to disallow the combination of spot fleet + awsNodeLabels.

**Comment by paalkr** [comment]
@mumoshu ,

Thanks

So I have a syntax error in the pull command. This is from the log file
I missed the ; after the latest done command. I'll give it another shot right now.
```
May 17 04:43:33 ip-10-1-44-242.eu-west-1.compute.internal bash[1208]: /bin/bash: -c: line 0: syntax error near unexpected token `echo'
May 17 04:43:33 ip-10-1-44-242.eu-west-1.compute.internal bash[1208]: /bin/bash: -c: line 0: `for img in xxxxxxxxxx.dkr.ecr.eu-west-1.amazonaws.com/ags/105:base_wa_ags_1.2.1; do echo pulling \$img; until /usr/bin/docker pull \$img; do echo retrying...; sleep 1; done; done echo successfully pulled all the images'
May 17 04:43:33 ip-10-1-44-242.eu-west-1.compute.internal systemd[1]: prepull.service: Control process exited, code=exited status=1
```

> Thanks for the good catch!
> Probably we're missing a validation to disallow the combination of spot fleet + awsNodeLabels.

Thanks. You are probably right. I haven't seen any problems as of yet caused by these small issue though.

**Comment by mumoshu** [comment]
@paalkr I'm looking forward to your result 👍 

> I haven't seen any problems as of yet caused by these small issue though.

Good to know 👍 
I'd appreciate it if you could submit an another github issue to track that anyway.

**Comment by paalkr** [comment]
@mumoshu ,

More syntax errors it seems, any suggestions? I must admit that Linux is not my native "language" ;)
The  image is valid, and I just xxx-ed out our AWS account part of the repo for the sake of posting this comment.

Looks like the img variable is never populated with the image reference, I'm I right?

```
May 17 05:21:56 ip-10-1-43-176.eu-west-1.compute.internal systemd[1]: /etc/systemd/system/prepull.service:3: Unknown lvalue 'RequiredBy' in section 'Unit'
May 17 05:21:56 ip-10-1-43-176.eu-west-1.compute.internal systemd[1]: /etc/systemd/system/prepull.service:7: Ignoring unknown escape sequences: "for img in xxxxx.dkr.ecr.eu-west-1.amazonaws.com/ags/105:base_wa_ags_1.2.1; do echo pulling \$img; until /usr/bin/docker pull \$img; do echo retrying...; sleep 1; done; done; echo successfully pulled all the images"
May 17 05:21:56 ip-10-1-43-176.eu-west-1.compute.internal systemd[1]: Starting Docker images pre-puller...
May 17 05:21:56 ip-10-1-43-176.eu-west-1.compute.internal bash[1313]: pulling $img
May 17 05:22:01 ip-10-1-43-176.eu-west-1.compute.internal bash[1313]: Error parsing reference: "$img" is not a valid repository/tag
May 17 05:22:01 ip-10-1-43-176.eu-west-1.compute.internal bash[1313]: retrying...
May 17 05:22:02 ip-10-1-43-176.eu-west-1.compute.internal bash[1313]: Error parsing reference: "$img" is not a valid repository/tag
```

**Comment by mumoshu** [comment]
@paalkr Could you try replacing all the `\$img` to just `$img`?

**Comment by paalkr** [comment]
@mumoshu 

The command works, but unfortunately authentication doesn't
`May 17 06:11:54 ip-10-1-43-19.eu-west-1.compute.internal bash[1224]: unauthorized: authentication required`

I was afraid this would be the result. I guess that the command is not picking up the ECR secret, wich of course make sense. Do we need to pull using rkt or something?

**Comment by mumoshu** [comment]
@paalkr If I remember correctly, in a k8s cluster, k8s automatically obtains docker-login creds from ECR for you. In this case though, we'd need to obtain creds ourselves.
Could you run something like `docker_login=$(docker run coreos/awscli aws ecr get-login); $docker_login` in the beginning of the script to run an appropriate `docker login`?

**Comment by paalkr** [comment]
@redbaron , @mumoshu 

Thanks a lot for all the good help and suggestions. I got the prepull service to work by adding this to my cluster config. Nice :)
```
      customSystemdUnits:
      - name: prepull.service
        command: start
        enable: true
        content: |
          [Unit]
          Description=Docker images pre-puller
          RequiredBy=kubelet.service
          [Service]
          Type=oneshot
          RemainAfterExit=yes
          ExecStartPre=/bin/bash -c "docker_login=$(docker run quay.io/coreos/awscli:master aws ecr get-login --region eu-west-1); $docker_login; for img in XXXXXXXX.dkr.ecr.eu-west-1.amazonaws.com/ags/105:base_wa_ags_1.2.1; do echo pulling $img; until /usr/bin/docker pull $img; do echo retrying...; sleep 1; done; done; echo successfully pulled all the images"
          ExecStart=/bin/true
```
